File: rplugin/node/vim-be-good/menu.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
class Menu {
    constructor(plugin, games, difficulties, selectedDifficulty, __allowCreation) {
        this.headerLines = [
            "VimBeGood2 is a collection of small games for neovim which are",
            "intended to help you improve your vim proficiency.",
            ""
        ];
        this.gameInstructions = [
            "Select a Game (delete from the list to select)",
            "----------------------------------------------"
        ];
        this.difficultyInstructions = [
            "",
            "Select a Difficulty (delete from the list to select)",
            "----------------------------------------------------"
        ];
        this.difficultyLines = [];
        this.footer = [
            "",
            "",
            "",
            "",
            "Created by ThePrimeagen",
            "           Brandoncc",
            "https://github.com/ThePrimeagen/vim-be-good"
        ];
        this.fullMenu = [];
        //eslint-disable-next-line @typescript-eslint/no-inferrable-types
        this.firstGameLineIndex = -1;
        //eslint-disable-next-line @typescript-eslint/no-inferrable-types
        this.firstDifficultyLineIndex = -1;
        if (!__allowCreation) {
            throw new Error("Menu cannot be instantiated, you must use the builder");
        }
        this.plugin = plugin;
        this.lineEventHandler = this.lineEventHandler.bind(this);
        this.gameList = games;
        this.difficultyList = difficulties;
        this.selectedDifficulty = difficulties[0];
        this.generateMenuLines();
    }
    static build(plugin, availableGames, availableDifficulties, selectedDifficulty) {
        return __awaiter(this, void 0, void 0, function* () {
            const menu = new Menu(plugin, availableGames, availableDifficulties, selectedDifficulty, true);
            yield menu.setup();
            return menu;
        });
    }
    setup() {
        return __awaiter(this, void 0, void 0, function* () {
            console.log("menu -- Setup Starting");
            this.buffer = yield this.plugin.nvim.buffer;
            this.window = yield this.plugin.nvim.window;
            console.log("menu -- Setup Finished");
        });
    }
    clearScreen() {
        var _a, _b;
        return __awaiter(this, void 0, void 0, function* () {
            console.log("menu -- clearScreen");
            yield ((_a = this.buffer) === null || _a === void 0 ? void 0 : _a.remove(0, yield ((_b = this.buffer) === null || _b === void 0 ? void 0 : _b.length), true));
        });
    }
    render() {
        var _a;
        return __awaiter(this, void 0, void 0, function* () {
            if (!this.window) {
                return;
            }
            this.stopHandlingLineEvents();
            yield this.clearScreen();
            yield ((_a = this.buffer) === null || _a === void 0 ? void 0 : _a.setLines(this.fullMenu, { start: 0 }));
            this.window.cursor = [this.firstGameLineIndex, 0];
            this.handleLineEvents();
        });
    }
    onGameSelection(cb) {
        this.gameSelectionCallback = cb;
    }
    lineEventHandler(buffer, _changedTick, _firstLine, lastLine, _newData, _more) {
        var _a;
        return __awaiter(this, void 0, void 0, function* () {
            try {
                if (!this.window) {
                    return;
                }
                const newLines = yield buffer.lines;
                const deletedCount = this.fullMenu.length - newLines.length;
                console.log("menu -- onLineChange", deletedCount);
                if (deletedCount === 1) {
                    const selectedGame = this.gameList[lastLine - this.firstGameLineIndex];
                    const selectedDifficulty = this.difficultyList[lastLine - this.firstDifficultyLineIndex];
                    console.log("menu -- onLineChange, if deletedCount === 1", selectedGame, selectedDifficulty);
                    if (selectedGame) {
                        console.log("menu -- onLineChange, selectedGame");
                        this.stopHandlingLineEvents();
                        yield this.clearScreen();
                        (_a = this.gameSelectionCallback) === null || _a === void 0 ? void 0 : _a.call(this, selectedGame, this.selectedDifficulty);
                    }
                    else if (selectedDifficulty) {
                        console.log("menu -- onLineChange, selectedDifficulty");
                        this.selectedDifficulty = selectedDifficulty;
                        this.stopHandlingLineEvents();
                        this.generateMenuLines();
                        yield this.render();
                    }
                    else {
                        console.log("menu -- onLineChange, NOTHING WAS FOUND");
                        yield this.render();
                    }
                }
                else {
                    yield this.render();
                }
            }
            catch (e) {
                yield this.plugin.nvim.outWrite("Error while handling line: " + e.message + "\n");
            }
        });
    }
    handleLineEvents() {
        var _a;
        (_a = this.buffer) === null || _a === void 0 ? void 0 : _a.listen("lines", this.lineEventHandler);
    }
    stopHandlingLineEvents() {
        var _a;
        (_a = this.buffer) === null || _a === void 0 ? void 0 : _a.unlisten("lines", this.lineEventHandler);
    }
    createDifficultyLines(difficulties) {
        return difficulties.map(diff => {
            if (this.selectedDifficulty === diff) {
                return `[X] ${diff}`;
            }
            else {
                return `[ ] ${diff}`;
            }
        });
    }
    generateMenuLines() {
        this.difficultyLines = this.createDifficultyLines(this.difficultyList);
        this.fullMenu = [
            ...this.headerLines,
            ...this.gameInstructions,
            ...this.gameList,
            ...this.difficultyInstructions,
            ...this.difficultyLines,
            ...this.footer
        ];
        this.firstGameLineIndex =
            this.headerLines.length + this.gameInstructions.length + 1;
        this.firstDifficultyLineIndex =
            this.headerLines.length +
                this.gameInstructions.length +
                this.gameList.length +
                this.difficultyInstructions.length +
                1;
    }
}
exports.Menu = Menu;

File: lua/vim-be-good/menu.lua
local log = require("vim-be-good.log")
local bind = require("vim-be-good.bind")
local types = require("vim-be-good.types")
local createEmpty = require("vim-be-good.game-utils").createEmpty

local Menu = {}

local gameHeader = {
    "",
    "Select a Game (delete from the list to select)",
    "----------------------------------------------",
}

local difficultyHeader = {
    "",
    "Select a Difficulty (delete from the list to select)",
    "Noob difficulty is endless so it must be quit with :q",
    "----------------------------------------------------",
}

local instructions = {
    "VimBeGood is a collection of small games for neovim which are",
    "intended to help you improve your vim proficiency.",
    "delete a line to select the line.  If you delete a difficulty,",
    "it will select that difficulty, but if you delete a game it ",
    "will start the game."
}

local credits = {
    "",
    "",
    "Created by ThePrimeagen",
    "Brandoncc",
    "polarmutex",
    "",
    "https://github.com/ThePrimeagen/vim-be-good",
    "https://twitch.tv/ThePrimeagen",
}

function Menu:new(window, onResults)
    local menuObj = {
        window = window,
        buffer = window.buffer,
        onResults = onResults,

        -- easy
        difficulty = types.difficulty[2],

        -- relative
        game = types.games[1],
    }

    window.buffer:clear()
    window.buffer:setInstructions(instructions)

    self.__index = self
    local createdMenu = setmetatable(menuObj, self)

    createdMenu._onChange = bind(createdMenu, "onChange")
    window.buffer:onChange(createdMenu._onChange)

    return createdMenu
end

local function getMenuLength()
    return #types.games + #types.difficulty + #gameHeader +
        #difficultyHeader + #credits
end

local function getTableChanges(lines, compareSet, startIdx)
    local maxCount = #lines
    local idx = startIdx
    local i = 1
    local found = false

    while found == false and idx <= maxCount and i <= #compareSet do
        if lines[idx] == nil or lines[idx]:find(compareSet[i], 1, true) == nil then
            found = true
        else
            i = i + 1
            idx = idx + 1
        end
    end

    return found, i, idx
end

function Menu:onChange()
    local lines = self.window.buffer:getGameLines()
    local maxCount = getMenuLength()

    if #lines == maxCount then
        return
    end

    local found, i, idx = getTableChanges(lines, gameHeader, 1)
    log.info("Menu:onChange initial instructions", found, i, idx)
    if found then
        self:render()
        return
    end

    found, i, idx = getTableChanges(lines, types.games, idx)
    log.info("Menu:onChange game changes", found, i, idx)
    if found then
        self.game = types.games[i]

        log.info("Starting Game", self.game, self.difficulty)
        ok, msg = pcall(self.onResults, self.game, self.difficulty)

        if not ok then
            log.info("Menu:onChange error", msg)
        end
        return
    end

    found, i, idx = getTableChanges(lines, difficultyHeader, idx)
    if found then
        self:render()
        return
    end

    found, i, idx = getTableChanges(lines, types.difficulty, idx)
    if found then
        self.difficulty = types.difficulty[i]
        self:render()
        return
    end
end

local function createMenuItem(str, currentValue)
    if currentValue == str then
        return "[x] " .. str
    end
    return "[ ] " .. str
end

function Menu:render()
    self.window.buffer:clearGameLines()

    local lines = { }
    for idx = 1, #gameHeader do
        table.insert(lines, gameHeader[idx])
    end

    for idx = 1, #types.games do
        table.insert(lines, createMenuItem(types.games[idx], self.game))
    end

    for idx = 1, #difficultyHeader do
        table.insert(lines, difficultyHeader[idx])
    end

    for idx = 1, #types.difficulty do
        table.insert(lines, createMenuItem(types.difficulty[idx], self.difficulty))
    end

    for idx = 1, #credits do
        table.insert(lines, credits[idx])
    end

    self.window.buffer:render(lines)
end

function Menu:close()
    self.buffer:removeListener(self._onChange)
end

return Menu

